tls/s2n_client_hello.c

```diff
@@ -310,6 +322,70 @@ static S2N_RESULT s2n_client_hello_verify_for_retry(struct s2n_connection *conn,
     return S2N_RESULT_OK;
 }
 
+S2N_RESULT s2n_client_hello_parse_raw(struct s2n_client_hello *client_hello,
+        uint8_t client_protocol_version[S2N_TLS_PROTOCOL_VERSION_LEN],
+        uint8_t client_random[S2N_TLS_RANDOM_DATA_LEN])
+{
+    RESULT_ENSURE_REF(client_hello);
+
+    struct s2n_stuffer in_stuffer = { 0 };
+    RESULT_GUARD_POSIX(s2n_stuffer_init(&in_stuffer, &client_hello->raw_message));
+    RESULT_GUARD_POSIX(s2n_stuffer_skip_write(&in_stuffer, client_hello->raw_message.size));
+    struct s2n_stuffer *in = &in_stuffer;
+
+    /**
+     * https://tools.ietf.org/rfc/rfc8446#4.1.2
+     * Structure of this message:
+     *
+     *    uint16 ProtocolVersion;
+     *    opaque Random[32];
+     *
+     *    uint8 CipherSuite[2];
+     *
+     *    struct {
+     *        ProtocolVersion legacy_version = 0x0303;
+     *        Random random;
+     *        opaque legacy_session_id<0..32>;
+     *        CipherSuite cipher_suites<2..2^16-2>;
+     *        opaque legacy_compression_methods<1..2^8-1>;
+     *        Extension extensions<8..2^16-1>;
+     *    } ClientHello;
+     **/
+
+    /* legacy_version */
+    RESULT_GUARD_POSIX(s2n_stuffer_read_bytes(in, client_protocol_version, S2N_TLS_PROTOCOL_VERSION_LEN));
+
+    /* random */
+    RESULT_GUARD_POSIX(s2n_stuffer_erase_and_read_bytes(in, client_random, S2N_TLS_RANDOM_DATA_LEN));
```

Comment:
Good question, it scared me for a minute ;) No, because we've already copied the customer's client hello into the s2n_client_hello's raw message blob. We just erase the random bytes from our copy.

I could make the s2n_client_hello_parse_message's `bytes` parameter `const`, but that would require us to copy it a second time in order to parse it with any of our stuffer or blob functions, because blobs and stuffers aren't const.

Reply:
I got a second opinion from @alexw91 and decided to make the function signature correct and use const. This isn't a super performance sensitive API, so the extra copy shouldn't be a problem.